scenes/cslib_string.txt

```diff
@@ -278,4 +278,39 @@
 	*set char_list_index char_list_len
 	*goto contains_loop
 *set cslib_ret false
-*return
\ No newline at end of file
+*return
+
+
+*comment CONSISTS_OF
+*comment ------------------------------
+*comment Checks if the given string consists
+*comment only of characters in a given list.
+*comment ------------------------------
+*comment 		params:
+*comment 			p_string(string): the string to check for the presence of the given cheracters. 
+*comment 			p_char_list(string): a string of characters allowed in the given string
+*comment 		returns:
+*comment 			(boolean): true if p_string consists only of chars in p_char_list, else false
+*label consists_of
+*params p_string p_char_list
+*set cslib_ret true
+*if (p_char_list = "")
+	*bug cslib_string.consists_of: Character list must not be empty.
+*if (p_string = "")
+	*set cslib_ret false
+	*return
+*temp p_string_len length(p_string)
+*temp p_char_list_len length(p_char_list)
+*temp p_string_index 1
+*label consists_of_string_loop
+*temp p_char_list_index 1
+*if (p_string_index <= p_string_len)
+	*label consists_of_char_list_loop
+	*if (p_char_list_index <= p_char_list_len)
+		*if ((p_string#p_string_index) = (p_char_list#p_char_list_index))
+			*set p_string_index +1
+			*goto consists_of_string_loop
+		*set p_char_list_index +1
+		*goto consists_of_char_list_loop
+	*set cslib_ret false
```

Comment:
you can also immediately return (after false)

Reply:
Not really. In the last iteration, when the if statement evaluates to false the routine would fall into nothing. The return statement must be on the same indentation level.

Reply:
Is that with implicit_control_flow set to true? (which the library actually requires atm btw).

Reply:
I made it agnostic. I suspected you'd be using implicit control, but I didn't want to rely on it. Anyway, I mean the loop repeats until the walker/index is greater than the string length. So we need a return after the if.

```
LABEL CONSISTS_OF
...
REPEAT AS LONG AS THERE'S STRING TO PARSE <- this will evaluate to false in the last iteration
	...
RETURN <- we need the return here
```

Reply:
what I meant is that you could slightly reorganize the code so that, as soon as one char from the string is NOT in your set, you update the result to false and return right there - so you don't have to always scan all the way to the end

BTW, it's really hard to read a CS loop  :-)

Reply:
🤔 I don't follow. The only string it follows to the end it the charset. If it gets to the end of the charset without any match, then it does return from there.

>  it's really hard to read a CS loop

I agree with reading the loops. We could add comments indicating the beginning and the end of the loop.

```
*comment BEGIN LOOP
*label consists_of_loop
...
*comment END LOOP 
```

Reply:
That's fine, as far as it works.